python/src/iceberg/expression/base.py

```diff
@@ -0,0 +1,146 @@
+        if rest:
+            return reduce(And, (left, right, *rest))
+        if left is alwaysFalse() or right is alwaysFalse():
+            return alwaysFalse()
+        elif left is alwaysTrue():
+            return right
+        elif right is alwaysTrue():
+            return left
+        return super().__new__(cls)
+
+    def __init__(self, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression):
+        if not rest:
+            self.left = left
+            self.right = right
+
+    def __eq__(self, other) -> bool:
+        return id(self) == id(other) or (isinstance(other, And) and self.left == other.left and self.right == other.right)
+
+    def __invert__(self) -> "Or":
+        return Or(~self.left, ~self.right)
+
+    def __repr__(self) -> str:
+        return f"And({repr(self.left)}, {repr(self.right)})"
+
+    def __str__(self) -> str:
+        return f"({self.left} and {self.right})"
+
+
+class Or(BooleanExpression):
+    """OR operation expression - logical disjunction"""
+
+    def __new__(cls, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression):
+        if rest:
+            return reduce(Or, (left, right, *rest))
+        if left is alwaysTrue() or right is alwaysTrue():
+            return alwaysTrue()
+        elif left is alwaysFalse():
+            return right
+        elif right is alwaysFalse():
+            return left
+        return super().__new__(cls)
+
+    def __init__(self, left: BooleanExpression, right: BooleanExpression, *rest: BooleanExpression):
+        if not rest:
+            self.left = left
+            self.right = right
+
+    def __eq__(self, other) -> bool:
+        return id(self) == id(other) or (isinstance(other, Or) and self.left == other.left and self.right == other.right)
+
+    def __invert__(self) -> "And":
+        return And(~self.left, ~self.right)
+
+    def __repr__(self) -> str:
+        return f"Or({repr(self.left)}, {repr(self.right)})"
+
+    def __str__(self) -> str:
+        return f"({self.left} or {self.right})"
+
+
+class Not(BooleanExpression):
+    """NOT operation expression - logical negation"""
+
+    def __new__(cls, child: BooleanExpression):
+        if child is alwaysTrue():
+            return alwaysFalse()
+        elif child is alwaysFalse():
+            return alwaysTrue()
+        elif isinstance(child, Not):
+            return child.child
+        return super().__new__(cls)
+
+    def __init__(self, child):
+        self.child = child
+
+    def __eq__(self, other) -> bool:
+        return id(self) == id(other) or (isinstance(other, Not) and self.child == other.child)
+
+    def __invert__(self) -> BooleanExpression:
+        return self.child
+
+    def __repr__(self) -> str:
+        return f"Not({repr(self.child)})"
+
+    def __str__(self) -> str:
+        return f"(not {self.child})"
+
+
+class alwaysTrue(BooleanExpression, Singleton):
```

Comment:
Should these be caplitalized since they are actually classes? That would be more consistent with how this is going to be used, since I think this combines the factory methods and classes by overriding `__new__` right?